klippa/src/cmap.rs

```diff
@@ -0,0 +1,1105 @@
+    drop_format_4: bool,
+) -> Result<(), SerializeErrorFlags> {
+    // allocate header: version + numTables
+    s.allocate_size(4, false)?;
+
+    let snap = s.snapshot();
+    let mut format12_objidx = None;
+    //TODO: add support for cmap_cache in plan accelerator
+    let mut unicodes_cache =
+        SubtableUnicodeCache::new(cmap.offset_data().as_bytes().as_ptr() as usize);
+    for (rec_idx, record) in retained_encoding_records.iter() {
+        if s.in_error() {
+            return Err(s.error());
+        }
+
+        let Ok(subtable) = record.subtable(cmap.offset_data()) else {
+            continue;
+        };
+
+        let format = subtable.format();
+        if format != 4 && format != 12 && format != 14 {
+            continue;
+        }
+
+        if !drop_format_4 && format == 4 {
+            let Some(unicodes_set) =
+                unicodes_cache.set_for(*rec_idx, &subtable, plan.font_num_glyphs)
+            else {
+                continue;
+            };
+            let cp_to_new_gid_list: Vec<(u32, GlyphId)> = plan
+                .unicode_to_new_gid_list
+                .iter()
+                .filter_map(|(cp, gid)| unicodes_set.contains(*cp).then_some((*cp, *gid)))
+                .collect();
+
+            serialize_encoding_record(record, &subtable, s, &cp_to_new_gid_list, plan)?;
+            if s.in_error() && s.only_overflow() {
+                // cmap4 overflowed, reset and retry serialization without format 4 subtables.
+                s.revert_snapshot(snap);
+                return serialize_cmap(cmap, s, plan, retained_encoding_records, true);
+            }
+        } else if format == 12 {
+            let Some(unicodes_set) =
+                unicodes_cache.set_for(*rec_idx, &subtable, plan.font_num_glyphs)
+            else {
+                continue;
+            };
+            let cmap12_subset_unicodes =
+                IntSet::from_iter(plan.unicodes.iter().filter(|v| unicodes_set.contains(*v)));
+
+            if can_drop_format12(
+                record,
+                &cmap12_subset_unicodes,
+                cmap,
+                retained_encoding_records,
+                &mut unicodes_cache,
+                &plan.unicodes,
+                plan.font_num_glyphs,
+            ) {
+                continue;
+            }
+
+            let cp_to_new_gid_list: Vec<(u32, GlyphId)> = plan
+                .unicode_to_new_gid_list
+                .iter()
+                .filter(|&(cp, _gid)| cmap12_subset_unicodes.contains(*cp))
+                .map(|(cp, gid)| (*cp, *gid))
+                .collect();
+            format12_objidx =
+                serialize_encoding_record(record, &subtable, s, &cp_to_new_gid_list, plan)?;
+        } else if format == 14 {
+            serialize_encoding_record(record, &subtable, s, &plan.unicode_to_new_gid_list, plan)?;
+        }
+    }
+
+    let num_retained_records = (s.length() - 4) / EncodingRecord::RAW_BYTE_LEN;
```

Comment:
I assume `4` is the header length? Might be worth a `const`, I think you use it elsewhere too and a const with a helpful name is more robust documentation than a comment.